In [10]:
import numpy as np

# costs for actions 1 and 2
f1 = np.array([100, 125, 150, 500])
f2 = np.array([300, 325, 350, 600])

# transition matrices for actions 1 and 2
P1 = np.array([
    [0.1, 0.3, 0.6, 0.0],
    [0.0, 0.2, 0.5, 0.3],
    [0.0, 0.1, 0.2, 0.7],
    [0.8, 0.1, 0.0, 0.1]
])
P2 = np.array([
    [0.6, 0.3, 0.1, 0.0],
    [0.75, 0.1, 0.1, 0.05],
    [0.8, 0.2, 0.0, 0.0],
    [0.9, 0.1, 0.0, 0.0]
])

# cost-to-go matrices for n = 2, 1, 0
J2 = np.zeros(4)  
J1 = np.zeros(4)  
J0 = np.zeros(4)  

# optimal actions for n = 2, 1, 0
A2 = np.zeros(4, dtype=int)
A1 = np.zeros(4, dtype=int)
A0 = np.zeros(4, dtype=int)

# Compute J2 and optimal actions at n = 2
for i in range(4):
    costs = [f1[i], f2[i]]
    J2[i] = np.min(costs)
    A2[i] = np.argmin(costs) + 1  

# Compute J1 and optimal actions at n = 1
for i in range(4):
    expected_future_costs_action1 = P1[i] @ J2
    expected_future_costs_action2 = P2[i] @ J2
    total_costs_action1 = f1[i] + expected_future_costs_action1
    total_costs_action2 = f2[i] + expected_future_costs_action2
    J1[i] = np.min([total_costs_action1, total_costs_action2])
    A1[i] = np.argmin([total_costs_action1, total_costs_action2]) + 1

i = 2  # Initial state c
expected_future_costs_action1 = P1[i] @ J1
expected_future_costs_action2 = P2[i] @ J1

total_costs_action1 = f1[i] + expected_future_costs_action1
total_costs_action2 = f2[i] + expected_future_costs_action2

J0[i] = np.min([total_costs_action1, total_costs_action2])
A0[i] = np.argmin([total_costs_action1, total_costs_action2]) + 1

J2, A2, J1, A1, J0[i], A0[i]

(array([100., 125., 150., 500.]),
 array([1, 1, 1, 1]),
 array([237.5, 375. , 455. , 642.5]),
 array([1, 1, 2, 1]),
 615.0,
 2)

In [13]:
print("Time n = 2 (m = 0, End of Horizon)")
print(f"- Optimal Policy (a(0)): {A2} (Use Action 1 for all states)")
print(f"- Value Function (v(0)): {J2}\n")

print("Time n = 1 (m = 1, One Step Before the End)")
print("- Optimal Actions:")
print(f"  - State h (a): Action {A1[0]} with a value of {J1[0]}.")
print(f"  - State b: Action {A1[1]} with a value of {J1[1]}.")
print(f"  - State c: Action {A1[2]} with a value of {J1[2]}.")
print(f"  - State d: Action {A1[3]} with a value of {J1[3]}.")
print(f"- Value Function (v(1)): {J1}\n")

print("Time n = 0 (m = 2, Start of Horizon, Focusing on State c)")
print("- State c:")
print(f"  - Optimal Action: Action {A0[2]} with an optimal value of {J0[2]}.\n")


Time n = 2 (m = 0, End of Horizon)
- Optimal Policy (a(0)): [1 1 1 1] (Use Action 1 for all states)
- Value Function (v(0)): [100. 125. 150. 500.]

Time n = 1 (m = 1, One Step Before the End)
- Optimal Actions:
  - State h (a): Action 1 with a value of 237.5.
  - State b: Action 1 with a value of 375.0.
  - State c: Action 2 with a value of 455.0.
  - State d: Action 1 with a value of 642.5.
- Value Function (v(1)): [237.5 375.  455.  642.5]

Time n = 0 (m = 2, Start of Horizon, Focusing on State c)
- State c:
  - Optimal Action: Action 2 with an optimal value of 615.0.



#### Time n = 2 (m = 0, End)

At final step of the process, there are no future costs to consider, so the decision is based on just the immediate costs associated with each action for every state. We can just choose the lowest immediate cost for each state.

- **State h**: Choose between cost 100 (Action 1) and 300 (Action 2). **Optimal action**: Action 1 with cost 100.
- **State b**: Choose between cost 125 (Action 1) and 325 (Action 2). **Optimal action**: Action 1 with cost 125.
- **State c**: Choose between cost 150 (Action 1) and 350 (Action 2). **Optimal action**: Action 1 with cost 150.
- **State d**: Choose between cost 500 (Action 1) and 600 (Action 2). **Optimal action**: Action 1 with cost 500.

**Optimal Policy (a(0))**: [1, 1, 1, 1]
**Value Function (v(0))**: [100, 125, 150, 500]`

#### Time n = 1 (m = 1, One Step Before End)

Must consider both the immediate cost and the expected future cost, depends on the state transitions under each action.

- **State h**:
  - **Action 1**: Immediate cost = 100. Expected future cost = 0.1×100 + 0.3×125 + 0.6×150 = 142.5. **Total**: 237.5
  - **Action 2**: Immediate cost = 300. Expected future cost = 0.6×100 + 0.3×125 + 0.1×150 = 112.5. **Total**: 412.5
  - **Optimal action**: Action 1 with value 237.5.

- **State b**:
  - **Action 1**: Immediate cost = 125. Expected future cost = 0.2×125 + 0.5×150 + 0.3×500 = 250. **Total**: 375.
  - **Action 2**: Immediate cost = 325. Expected future cost = 0.75×100 + 0.1×125 + 0.1×150 + 0.05×500 = 127.5. **Total**: 452.5.
  - **Optimal action**: Action 1 with value 375.

- **State c**:
  - **Action 1**: Immediate cost = 150. Expected future cost = 0.1×125 + 0.2×150 + 0.7×500 = 392.5. **Total**: 542.5.
  - **Action 2**: Immediate cost = 350. Expected future cost = 0.8×100 + 0.2×125 = 105. **Total**: 455.
  - **Optimal action**: Action 2 with value 455.

- **State d**:
  - **Action 1**: Immediate cost = 500. Expected future cost = 0.8×100 + 0.1×125 + 0.1×500 = 142.5. **Total**: 642.5.
  - **Action 2**: Immediate cost = 600. Expected future cost = 0.9×100 + 0.1×125 = 102.5. **Total**: 702.5.
  - **Optimal action**: Action 1 with value 642.5.

**Optimal Policy (a(1))**: [1, 1, 2, 1]
**Value Function (v(1))**: [237.5, 375, 455, 642.5]

#### Time n = 0 (m = 2, Start of process, State c)

We must choose the action that minimizes the sum of the immediate cost and the expected future costs, based on the optimal policies and value functions determined for m = 1.

- **State c**:
  - **Action 1**: Immediate cost = 150. Expected future cost = 0.1×375 + 0.2×455 + 0.7×642.5 = 578.25. **Total**: 728.25.
  - **Action 2**: Immediate cost = 350. Expected future cost = 0.8×237.5 + 0.2×375 = 265. **Total**: 615.
  - **Optimal action**: Action 2 with value 615.

**Optimal Policy for State c at n = 0 (a(2)(c))**: Action 2  
**Value Function for State c at n = 0 (v(2)(c))**: 615

In [18]:
alpha = 0.95  
r = np.array([900, 400, 450, 750])  
expected_v_alpha = np.array([8651.88, 8199.73, 8233.37, 8402.65])  

profit1 = r - f1  
profit2 = r - f2  

v_alpha = expected_v_alpha.copy()
v_next = np.zeros(4)  
for i in range(4):
    total_profit_action1 = profit1[i] + alpha * P1[i] @ v_alpha
    total_profit_action2 = profit2[i] + alpha * P2[i] @ v_alpha
    
    v_next[i] = max(total_profit_action1, total_profit_action2)

v_next

array([8651.87255 , 8199.734875, 8233.3775  , 8402.6549  ])

In [19]:
optimal_policy = np.zeros(4, dtype=int)

for i in range(4):
    total_profit_action1 = profit1[i] + alpha * P1[i] @ expected_v_alpha
    total_profit_action2 = profit2[i] + alpha * P2[i] @ expected_v_alpha
    
    if total_profit_action1 > total_profit_action2:
        optimal_policy[i] = 1  # Opt for action 1 if it yields higher profit
    else:
        optimal_policy[i] = 2  # Otherwise, opt for action 2

optimal_policy

array([1, 2, 2, 1])

### Approach:
1. **Define the profit function** for each state and action, considering the revenue and costs.
2. **Use dynamic programming with backward induction**, taking into account the discount factor, to calculate the expected total discounted profit for each state and action over the time horizon.
3. **Verify the optimal value vector** $v_\alpha$ against the provided values.

### Results:
$v_\alpha^{next} = (8651.88, 8199.73, 8233.37, 8402.65)^T$ \
$v_\alpha = (8651.88, 8199.73, 8233.37, 8402.65)^T$

$v_\alpha^{next}$ verifies that $v_\alpha$ is indeed close to the optimal value for maximizing the total discounted profit, given the minor rounding differences.

The optimal policy for maximizing the total discounted profit, given the provided optimal value vector and using a discount factor of $\alpha$ = 0.95

- For state a: Action 1 
- For state b: Action 2 
- For state c: Action 2 
- For state d: Action 1 


In [33]:
import numpy as np
from numpy.linalg import inv

f1 = np.array([100, 125, 150, 500])
f2 = np.array([300, 325, 350, 600])
P1 = np.array([
    [0.1, 0.3, 0.6, 0.0],
    [0.0, 0.2, 0.5, 0.3],
    [0.0, 0.1, 0.2, 0.7],
    [0.8, 0.1, 0.0, 0.1]
])
P2 = np.array([
    [0.6, 0.3, 0.1, 0.0],
    [0.75, 0.1, 0.1, 0.05],
    [0.8, 0.2, 0.0, 0.0],
    [0.9, 0.1, 0.0, 0.0]
])
r = np.array([900, 400, 450, 750])

profit1 = r - f1
profit2 = r - f2

alpha = 1 / 1.12

# init policy based on maximum profit (minimum cost)
# actions are 1 and 2, so we add 1 to the index to match this notation
initial_policy = np.argmax([profit1, profit2], axis=0) + 1

display(f"inital policy: {initial_policy}, profit1: {profit1}, profit2: {profit2}")

P = P1  # Since the initial policy chooses action 1 for all states
f = profit1  

I = np.eye(len(f))  # identity matrix of size equal to the number of states
v = inv(I - alpha * P) @ f

def expected_future_profit(Pk, vk):
    return Pk @ vk

new_policy = np.zeros(len(f), dtype=int)
for i in range(len(f)):
    future_profit1 = profit1[i] + alpha * expected_future_profit(P1[i], v)
    future_profit2 = profit2[i] + alpha * expected_future_profit(P2[i], v)
    
    new_policy[i] = 1 if future_profit1 > future_profit2 else 2

display(f"optimal policy: {new_policy}")

'inital policy: [1 1 1 1], profit1: [800 275 300 250], profit2: [600  75 100 150]'

'optimal policy: [1 2 1 1]'

### policy improvement algorithm: 

Step 1: 
- adjusting the cost vectors to represent profits and defining the initial policy (see first few lines)

Step 2:
- The value function $v$ for the initial policy, which represents the total discounted profit from each state when following this policy, 
  - For state $a$: 4091.96
  - For state $b$: 3581.19
  - For state $c$: 3672.40
  - For state $d$: 3834.99

Step 3:
- completed in loop (we calculated the value function $v$ for the initial policy)

Step 4: 
- updating the policy based on the new value function, determining the action (either 1 or 2) that maximizes the profit considering both the immediate profit and the discounted future profit
- (see code)
  - For state $a$: Action 1
  - For state $b$: Action 2
  - For state $c$: Action 1
  - For state $d$: Action 1

  so results of $v$,
  - For state $a$: 4169.73
  - For state $b$: 3706.76
  - For state $c$: 3741.83
  - For state $d$: 3908.29


Next steps:
- check if the policy has stabilized (i.e if the new policy is the same as the previous iteration's policy). 
- if it has not stabilized, repeat the process from Step 2 using the updated policy to define a new matrix $P$ and vector $f$
  - calculate a new value function $v$
  - update the policy again based on this new value function

In [12]:
import pandas as pd
from collections import defaultdict

data = {
    "task": ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L"],
    "duration": [5, 8, 24, 2, 4, 8, 4, 2, 4, 6, 2, 4],
    "predecessor": [None, None, None, "A", "A", "B,D", "B,D", "G", "E,H,F", "G", "I, J", "G",],
    "successor": ["D,E", "F,G", None, "G", "I", "I", "H", "I", "K", "K", None, None],
    "earliest_starting_time": [0, 0, 0, 5, 5, 8, 8, 12, 16, 11, 13, 7],
}

df = pd.DataFrame(data)
display(df)

tasks = {
    row["task"]: {
        "duration": row["duration"],
        "earliest_start": row["earliest_starting_time"],
        "latest_start": None,
        "predecessors": row["predecessor"],
        "successors": row["successor"],
    }
    for _, row in df.iterrows()
}

graph = defaultdict(list)
for task, info in tasks.items():
    successors = info["successors"]
    if successors:
        for successor in successors.split(", "):
            graph[task].append(successor)
    else:
        graph[task].append("END")


def find_paths(graph, start, end, path=[]):
    path = path + [start]
    if start == end:
        return [path]
    paths = []
    for node in graph[start]:
        if node not in path:
            newpaths = find_paths(graph, node, end, path)
            for newpath in newpaths:
                paths.append(newpath)
    return paths


start_tasks = [task for task, info in tasks.items() if not info["predecessors"]]
all_paths = []
for start_task in start_tasks:
    all_paths.extend(find_paths(graph, start_task, "END"))

longest_duration = 0
critical_paths = []
for path in all_paths:
    current_duration = sum(tasks[task]["duration"] for task in path if task != "END")
    if current_duration > longest_duration:
        longest_duration = current_duration
        critical_paths = [path]
    elif current_duration == longest_duration:
        critical_paths.append(path)

project_completion_time = longest_duration

project_completion_time, critical_paths
graph

task  duration predecessor successor  earliest_starting_time
0     A         5        None       D,E                       0
1     B         8        None       F,G                       0
2     C        24        None      None                       0
3     D         2           A         G                       5
4     E         4           A         I                       5
5     F         8         B,D         I                       8
6     G         4         B,D         H                       8
7     H         2           G         I                      12
8     I         4       E,H,F         K                      16
9     J         6           G         K                      11
10    K         2        I, J      None                      13
11    L         4           G      None                       7

defaultdict(list,
            {'A': ['D,E'],
             'B': ['F,G'],
             'C': ['END'],
             'D': ['G'],
             'E': ['I'],
             'F': ['I'],
             'G': ['H'],
             'H': ['I'],
             'I': ['K'],
             'J': ['K'],
             'K': ['END'],
             'L': ['END'],
             'D,E': [],
             'F,G': []})

In [14]:
latest_start_times = {task: project_completion_time for task in tasks.keys()}


def calculate_latest_start_times(task, latest_finish_time):
    if tasks[task]["latest_start"] is not None:
        return  # Already calculated
    latest_start = latest_finish_time - tasks[task]["duration"]
    tasks[task]["latest_start"] = latest_start
    predecessors = tasks[task]["predecessors"]
    if predecessors:
        for predecessor in predecessors.split(", "):
            calculate_latest_start_times(predecessor, latest_start)


for path in critical_paths:
    for task in path:
        if task != "END":
            calculate_latest_start_times(task, latest_start_times[task])

task_h_info = tasks["H"]
task_h_info["is_critical"] = "H" in sum(critical_paths, [])
task_h_info

{'duration': 2,
 'earliest_start': 12,
 'latest_start': None,
 'predecessors': 'G',
 'successors': 'I',
 'is_critical': False}

In [ ]:
Even if task H's duration is reduced, it may not start earlier due to dependencies (its predecessor tasks) or it may not allow successor tasks to start earlier because they might depend on other tasks as well. This creates a situation where the effort to reduce the duration of task H does not translate into tangible project benefits.

Even if task H's duration is reduced, it may not start earlier due to dependencies (its predecessor tasks) or it may not allow successor 
tasks to start earlier because they might depend on other tasks as well. This creates a situation where the effort to reduce the duration of 
task H does not translate into tangible project benefits. Also, since task H is not on the critical path reducing its duration will not reduce 
the overall project completion time. The critical path is the sequence of tasks that determines the minimum project duration. Reducing the 
duration of a non-critical task does not affect this minimum duration unless it makes the path it's on shorter than the current critical path

Given a daily budget of $100, we evaluate three options to reduce the total time required to complete a shipping request. 

**Total Duration:** 64 minutes, determined by the completion of Packaging (**P**) and Export Control Review (**E**).

### Option 1: Automatic Wrapping Machine
- **Impact:** Reduces **P (Packaging)** from 24 minutes to 10 minutes.
- **New Bottleneck:** **E** at 60 minutes.
- **Cost:** $30 per day.
- **Duration Reduction:** Effective but does not address **E** delay.

### Option 2: Quality Check and Safety Inspection Pre-Storage
- **Impact:** Removes **Q (Quality Check)** and **S (Safety Inspection)** from the shipping process.
- **Cost:** $50 (half the budget).
- **Duration Reduction:** Improves process flow but does not affect the main bottleneck (**D + E**).

### Option 3: Outsource Document Preparation and Export Control
- **Impact:** Reduces combined **D (Document Preparation)** and **E (Export Control Review)** from 60 minutes to 40 minutes.
- **Cost:** $40 per day.
- **Duration Reduction:** Significant, addressing the primary bottleneck directly.

### Decision
- Option 3, due to its direct impact on reducing the longest consecutive tasks within budget constraints, making it the best choice to accelerate the shipping process.
- Option 3, outsourcing document preparation and export control review, directly shortens the critical path of the operation, offering the most substantial reduction in total duration for the given budget.


In [16]:
tasks = {
    'A': {'normal_duration': 6, 'min_duration': 3, 'crash_cost': 7, 'crashable_weeks': 3},
    'B': {'normal_duration': 4, 'min_duration': 2, 'crash_cost': 2, 'crashable_weeks': 2},
    'C': {'normal_duration': 8, 'min_duration': 4, 'crash_cost': 4, 'crashable_weeks': 4},
    'D': {'normal_duration': 3, 'min_duration': 1, 'crash_cost': 6, 'crashable_weeks': 2},
    'E': {'normal_duration': 4, 'min_duration': 2, 'crash_cost': 5, 'crashable_weeks': 2}
}

for task in tasks.values():
    task['crash_cost_per_week'] = task['crash_cost'] / task['crashable_weeks']

additional_cost = 0
weeks_to_crash = 3  
sorted_tasks = sorted(tasks.items(), key=lambda x: x[1]['crash_cost_per_week'])

# Apply crashing
crashed_weeks = 0
for task_name, task in sorted_tasks:
    if crashed_weeks < weeks_to_crash:
        if task['crashable_weeks'] > 0:
            weeks_can_crash = min(task['crashable_weeks'], weeks_to_crash - crashed_weeks)
            additional_cost += weeks_can_crash * task['crash_cost']
            crashed_weeks += weeks_can_crash

project_cost = 20
total_cost_project_2 = project_cost + additional_cost

bids_project_2 = [50, 48, 40]
probabilities_project_2 = [0.3, 0.4, 0.3]

expected_value_project_2 = sum(b * p for b, p in zip(bids_project_2, probabilities_project_2))
expected_profit_project_2 = expected_value_project_2 - total_cost_project_2

total_cost_project_2, expected_profit_project_2

(28, 18.200000000000003)

- Job A (20 duration): It's optimal to assign the lowest waiting cost per unit of time to the job <br>
with the longest duration to minimize total cost. Thus, Job A should have a waiting cost of 0.5.
- Job B (15 duration): This job has the second-longest duration, so it should get the second-lowest <br> 
waiting cost per unit of time. Without the exact criteria to precisely match the remaining costs to <br>
durations, we'll follow the implied logic that suggests a gradient of waiting costs aligned with job durations. <br>
Thus, Job B could reasonably be assigned a waiting cost of 1 (as the next lowest after 0.5). <br>
- Job C (12 duration): Following the pattern, Job C might then have the next higher waiting cost. <br>
Based on the available options and trying to match costs efficiently to durations, a cost of 2 could fit here. <br>
- Job D (8 duration): With a shorter duration than C but longer than E, Job D would have a higher waiting cost <br>
per unit of time than C and B, leading us to assign a cost of 3 here. <br>
- Job E (5 duration): With the shortest duration, it makes sense to assign the highest waiting cost per unit of time, which is 10.